### ISSUE TITLE
Error occur after upgrading to twilio-video.js@2.0.0-beta6

### ISSUE THREAD

**Issue by smkrn110** [twilio/twilio-video.js#581]
Check the attached image for the error details:

![Screenshot from 2019-03-20 15-06-15](https://user-images.githubusercontent.com/2698458/54678533-23539a00-4b27-11e9-98c3-786c730ad986.png)

 - [x] I verified that the error is not occurs with previous version 2.0.0-beta5.
 - [x] I have verified that the issue occurs with the latest twilio-video.js release and is not marked as a known issue in the [CHANGELOG.md](https://github.com/twilio/twilio-video.js/blob/master/CHANGELOG.md).
 - [x] I reviewed the [Common Issues](https://github.com/twilio/twilio-video.js/blob/master/COMMON_ISSUES.md) and open GitHub issues and verified that this report represents a potentially new issue.

**Software versions:**

 - [x] Browser(s): Chrome
 - [x] Operating System: Ubuntu
 - [x] twilio-video.js: 2.0.0-beta6
 - [x] Third-party libraries (e.g., Angular, React, etc.): Angular 5

**Comment by manjeshbhargav** [comment]
Hi @smkrn110 ,

Thanks for writing in with your issue. Are you seeing this error during load time or after you have joined a Room? Also, Which version of Chrome did you use while you saw this error?

Thanks,

Manjesh Malavalli
JSDK Team

**Comment by smkrn110** [comment]
Hi @manjeshbhargav 

After joining of a room, then the library throwing this issue continuously.

 - [x] Browser(s): Chrome Version 71.0.3578.98 (Official Build) (64-bit)

**Comment by manjeshbhargav** [comment]
Hi @smkrn110 ,

I tried to reproduce this on my end and could not. Can you add a breakpoint on the line generating the TypeError and see if `response` is of type `RTCStatsReport`. Also, it would be great if you can share the output with me.

Thanks,

Manjesh Malavalli
JSDK Team

**Comment by smkrn110** [comment]
Attached is the screenshot with the breakpoint and on console, you can check. 

![Screenshot from 2019-03-25 15-01-46](https://user-images.githubusercontent.com/2698458/54911461-5e334480-4f10-11e9-907e-bf2e82cb8825.png)

I experienced the same issue today.

**Comment by manjeshbhargav** [comment]
Hi @smkrn110 ,

From your screenshot, it looks like `response` is an instance of the Chromium's legacy `RTCStatsResponse` object, which is only used in the SDK for Chrome <67. For Chrome 67+, `response` should be an instance of `RTCStatsReport`.  Can you also share the exact `userAgent` string with me? And is this Chrome browser or Electron?

Thanks,

Manjesh Malavalli
JSDK Team

**Comment by smkrn110** [comment]
```
navigator.userAgent

 "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36"
```
**Operating System:** Ubuntu 18.04.

**Comment by smkrn110** [comment]
@manjeshbhargav 

Any update on this issues, we already waited for beta6, for re-connection of signaling but encounter with the above issue.

2nd once we will move to beta6, and in a case when 'reconnecting' > 'reconnected' event fires, do the stream automatically recovers that is start playing (audio/video)? or we have to remove/unpublish tracks and republish them?

**Comment by manjeshbhargav** [comment]
Hi @smkrn110 ,

We are trying to reproduce this issue, and we will reach out once we have more information. Regarding your question about media during reconnection, it should automatically recover once you rejoin the Room. You won't need to unpublish and republish tracks.

Thanks,

Manjesh Malavalli
JSDK Team

**Comment by syerrapragada** [comment]
Hi @smkrn110,

We have not been able to reproduce the issue. Could you please run the following fiddle in the browser and paste the console output?

https://jsfiddle.net/f8vsw470/1/

Thanks,
Sreeram

**Comment by smkrn110** [comment]
@syerrapragada 

Following is the details, after executing the snippet you shared me via jsFiddle, also I have attached chrome about info with the version of the chrome. (OS : Ubuntu)*

![Screenshot from 2019-04-08 11-15-54](https://user-images.githubusercontent.com/2698458/55702385-e543ea80-59ef-11e9-8858-bb0e6b5595fc.png)
![Screenshot from 2019-04-08 11-14-15](https://user-images.githubusercontent.com/2698458/55702387-e543ea80-59ef-11e9-90fb-643e9ce14e93.png)

**Comment by smkrn110** [comment]
@syerrapragada 

After upgrade to chrome version 73, I am getting **RTCStatsReport**, 🧐

**Query** : If I upgrade to twilio version 2.0.0-beta6, so my application going to be unstable with previous version of chrome? 😕 😮

![Screenshot from 2019-04-08 11-51-56](https://user-images.githubusercontent.com/2698458/55704183-fa6f4800-59f4-11e9-8061-a0f82a5429a1.png)
![Screenshot from 2019-04-08 11-52-07](https://user-images.githubusercontent.com/2698458/55704184-fa6f4800-59f4-11e9-81a6-56566d6522bd.png)

**Comment by manjeshbhargav** [comment]
Hi @smkrn110 ,

Your application should still work without any issues. It is surprising that in Chrome 71 on Ubuntu, the Promise-based `RTCPeerConnection.getStats()` returns `RTCStatsResponse` which is deprecated. We will try and file a Chromium bug about this, although I'm not sure what their policy is about fixing bugs on older versions of Chrome (on Ubuntu in particular, this is not happening on Mac OS and Windows).

Thanks,

Manjesh Malavalli
JSDK Team

**Comment by smkrn110** [comment]
> Your application should still work without any issues. 

But the issue I reported will be face by a user having Chrome 71 with Ubuntu machine! It causing participants webcams are not showing either side. The main reason for reporting this, otherwise I would have put check around the `response.forEach` 🙂 and go with it.

I am totally confused and not sure to upgrade for **beta6** now, where Signaling re-connection mechanism is also important for me, any chance Twilio could help for this in next release?

**Comment by manjeshbhargav** [comment]
Hi @smkrn110 ,

We periodically publish webrtc stats to our servers so that we can debug any call quality issues later. I think this logic is what is resulting in the errors that you are seeing. A potential workaround can be to disable this feature on Chrome 71 like this:
```js
const room = await Twilio.Video.connect(token, {
  ...
  insights: isChrome71 ? false : true
  ...
});
```
Please try it our and let me know if this is good enough to unblock you.

Thanks,

Manjesh Malavalli
JSDK Team

**Comment by syerrapragada** [comment]
@smkrn110 
Did above workaround help mitigate the issue on Chrome 71 (Ubuntu)? Please let us know.

Thanks,
Sreeram

**Comment by smkrn110** [comment]
@manjeshbhargav or @syerrapragada 

Need quick response I am upgrading to beta11. The above solution not works with the provided snippet above for this issue.

```javascript
function doesGetStatsResolveWithRTCStatsResponse() {
  const pc = new RTCPeerConnection();
  const promise = pc.getStats();
  return promise instanceof Promise
    ? promise.then(stats => typeof RTCStatsResponse !== 'undefined'
      && stats instanceof RTCStatsResponse)
    : Promise.resolve(false);
}

doesGetStatsResolveWithRTCStatsResponse().then(doesResolve => {
  return Twilio.Video.connect(token, {
    ...
    insights: !doesResolve
    ...
  });
}).then(room => {
  ...
});
```

I have tried with removing the ! sign `insights: doesResolve` as doesGetStatsResolveWithRTCStatsResponse resolves to false, so I think ! sign is not needed. 

Can respond with quick fix for the issue!

Thanks

**Comment by smkrn110** [comment]
@syerrapragada 

Snippet you shared previously ..,

```javascript
var chromeMajorVersion = parseInt(navigator.userAgent.match(/Chrome\/([0-9]+)/)[1], 10);
console.log('Chrome major version:', chromeMajorVersion);

['plan-b', 'unified-plan'].forEach(sdpFormat => {
  new RTCPeerConnection({
    sdpSemantics: sdpFormat
  }).getStats().then(
    stats => console.log(`SDP Format:${sdpFormat}, stats object type: "${typeof stats}", has forEach?: ${typeof stats.forEach !== 'undefined'}`));
});
```

**Result**

![Screenshot from 2019-06-13 16-50-42](https://user-images.githubusercontent.com/2698458/59430444-b2f1a780-8dfb-11e9-99ac-6ee18629c001.png)

**Comment by smkrn110** [comment]
@manjeshbhargav 

![Screenshot from 2019-06-13 16-59-32](https://user-images.githubusercontent.com/2698458/59430842-ba658080-8dfc-11e9-9677-f2cbff11ab51.png)

![Screenshot from 2019-06-13 16-58-06](https://user-images.githubusercontent.com/2698458/59430839-b76a9000-8dfc-11e9-9731-154ff04907b3.png)

**Comment by smkrn110** [comment]
I experienced the same error on Windows, Chrome v74 and v75. forEach is not a function.  
```javascript
//twilio-video.js:20963
function standardizeChromeOrSafariStats(response) {
  // response.forEach
}
```

Kindly check and updated, following are the platform I have.

Windows: Chrome v74/75 
Ubuntu: Chrome v73

**Comment by manjeshbhargav** [comment]
Hi @smkrn110 ,

What do you get if you run `typeof RTCStatsResponse` in the browser console?

Thanks,

Manjesh Malavalli
JSDK Team

**Comment by syerrapragada** [comment]
Hi @smkrn110,

I tried on windows chrome v75 and I get the following:

![image](https://user-images.githubusercontent.com/20006575/59469565-04913680-8dea-11e9-9245-d69b3ee0bfa3.png)

Are you using any custom chrome build? 

Regards,
Sreeram

**Comment by smkrn110** [comment]
@syerrapragada 

Can you check this with chrome on ubuntu?

**Comment by smkrn110** [comment]
hi @manjeshbhargav & @syerrapragada 

ISSUE FIXED!

On drill down the **twilio-video.js** I encounter with WebRTC module which enlightened and assist me figuring out this issue. Actually there was a library **webrtc.js** which was previously used by the product currently I am working on, the main culprit behind the issue. It was overriding object RTCStatsResponse (and others) causing all that mess and on removing that webrtc.js, all working perfectly fine!

Closing this issue.

Regards,
